src/mbgl/renderer/renderer_impl.cpp

```diff
@@ -369,6 +369,47 @@ void Renderer::Impl::render(const UpdateParameters& updateParameters) {
         parameters.frameHistory.upload(parameters.context, 0);
     }
 
+    // - PREPARE + CLIP ------------------------------------------------------------------------------
+    {
+        MBGL_DEBUG_GROUP(parameters.context, "clip");
+
+        // Update all clipping IDs.
+        for (const auto& entry : renderSources) {
+            if (entry.second->isEnabled()) {
+                entry.second->startRender(parameters);
+            }
+        }
+    }
```

Comment:
Drawing the stencil mask before the 3D pass doesn't seem necessary since we're not using it, right? Drawing the clipping mask into the stencil buffer *before* means that we're drawing to the FBO *before* the clear step, which means that the OpenGL implementation has to restore the previous frame's stencil buffer (and it might potentially produce inaccurate clipping masks).

Reply:
Right — I didn't dig into `startRender` fully, but it looks like it both generates clipping IDs (I don't know how this works) _and_ [uploads buffers](https://github.com/mapbox/mapbox-gl-native/blob/f8a53234ffa6ff3fd1d8fc8e3aa8b10b16c815fd/src/mbgl/renderer/render_tile.cpp#L60). (Without moving this step before drawing, the fill-extrusion primitive drawing runs into `EXC_BAD_ACCESS` here: )
![image](https://user-images.githubusercontent.com/3170312/30218559-2aac3c1a-946e-11e7-9c31-45f69ef7de34.png)
And then like I said I know nothing about how the stencil drawing works but it at least _reads_ like the actual rendering of stencil masks is still done in a separate step: https://github.com/mapbox/mapbox-gl-native/blob/7531448275737ec473e6b69fcdb01911073cf1e0/src/mbgl/renderer/renderer_impl.cpp#L426-L429

If that's the case, maybe copying the descriptions of this (comments + debug group) exactly is misleading and should be updated to sound more like "upload buffers + generate clipping IDs". Or does uploading and generating clipping IDs need to be separated?

Reply:
Yeah, it looks like the code in the `CLIPPING MASKS` section does a few things: `startRender` collects the clipping IDs (CPU code only, no OpenGL involved here), and uploads buffers (OpenGL *is* involved here), then proceeds to actually render the clipping mask. I think we can split this up so we have the following code flow:
* Bind 
* Call `upload`/`startRender`
* 3D pass
* Clear

Looking at the current code, this is already how it's set up. I assumed that `startRender` would actually be drawing, but that turned out to be false.

Reply:
I consolidated this group with the "upload" group above it in https://github.com/mapbox/mapbox-gl-native/pull/9931/commits/9e786c19b496a4b94b0db2d06a8ef6be11ad07c1.